qemu/tests/thin_provisioning.py

```diff
@@ -0,0 +1,170 @@
+        scsi_debug = [_ for _ in scsi_disk_info][-1]
+    scsi_debug = scsi_debug.split()
+    scsi_id = scsi_debug[0][1:-1]
+    device_name = scsi_debug[-1]
+    return scsi_id, device_name
+
+
+def get_provisioning_mode(pci_id, device, session=None):
+    """
+    Get disk provisioning_mode, value usually is 'writesame_16', depends
+    on params for scsi_debug module.
+
+    :param pci_id: Pci slot number
+    :param device: The disk name
+    :param session: Guest session
+    :return:
+    """
+    device_name = os.path.basename(device)
+    path = "/sys/block/{:s}/device/scsi_disk".format(device_name)
+    path += "/{:s}/provisioning_mode".format(pci_id)
+    try:
+        if session is None:
+            return utils.read_one_line(path).strip()
+        else:
+            session.cmd("read < {:s}".format(path))
+            return session.cmd_output("echo $REPLY")
+    except IOError:
+        logging.error("Can't get the provisioning_mode", logging.error)
+
+
+def get_allocation_bitmap():
+    """
+    Get block allocation bitmap
+    """
+    path = "/sys/bus/pseudo/drivers/scsi_debug/map"
+    try:
+        bitmap = utils.read_one_line(path).strip()
+        return bitmap
+    except IOError:
+        raise exceptions.TestError("block allocation bitmap doesn't exists")
+
+
+def get_guest_discard_disk(session):
+    """
+    Get disk without partitions in guest.
+
+    :param session: Guest session
+    :return: Guest disk name
+    """
+    list_disk_cmd = "ls /dev/[shv]d*|sed 's/[0-9]//p'|uniq -u"
+    disk = session.cmd_output(list_disk_cmd).splitlines()[0]
+    return disk
+
+
+def destroy_vm(env):
+    for vm in env.get_all_vms():
+        if vm:
+            vm.destroy()
+            env.unregister_vm(vm.name)
+
+
+@error.context_aware
+def run(test, params, env):
+    """
+      'thin-provisioning' functions test using scsi_debug:
+      1) Boot up the guest
+      2) Create a file in the guest disk
+      3) Get the Bitmap and check it
+      4) Format the disk
+      5) Get the Bitmap and check it
+
+      :param test: QEMU test object
+      :param params: Dictionary with the test parameters
+      :param env: Dictionary with test environment.
+      """
+
+    # Destroy all vms to avoid emulated disk marked drity before start test
+    destroy_vm(env)
+    os_dep.command("lsscsi")
+    host_id, disk_name = get_scsi_disk()
+    provisioning_mode = get_provisioning_mode(host_id, disk_name)
+    logging.info("Scsi pci id and plot is: {:s} and disk name is: {:s}".format(host_id, disk_name))
+    logging.info("Current provisioning_mode = {:s}", provisioning_mode)
+
+    # Bitmap must be empty before test
+    if get_allocation_bitmap():
+        logging.debug("Block allocation bitmap: %s" % get_allocation_bitmap())
+        raise error.TestError("Block allocation bitmap not empty before test.")
+
+    vm_name = params["main_vm"]
+    test_image = params.get("test_image")
```

Comment:
For most cases, we use stg as image tag, so i think it will be consistent to use "stg" instead "scsi_debug". Also some configuration can be moved into cfg file like below:
image_format_stg = "raw"
image_raw_device_stg = "yes"
create_image_stg = "no"
force_create_image_stg = "no"
drv_extra_params_scsi_debug = "discard = on"
images += " stg"